packages/inter-protocol/src/auction/scheduler.js

```diff
@@ -0,0 +1,199 @@
+import { E } from '@endo/eventual-send';
+import { TimeMath } from '@agoric/time';
+import { Far } from '@endo/marshal';
+import { natSafeMath } from '@agoric/zoe/src/contractSupport/index.js';
+import { makeTracer } from '@agoric/internal';
+import { AuctionState } from './util.js';
+
+const { Fail } = assert;
+const { subtract, multiply, floorDivide } = natSafeMath;
+
+const trace = makeTracer('SCHED', false);
+
+/**
+ * @file The scheduler is presumed to be quiescent between auction rounds. Each
+ * Auction round consists of a sequence of steps with decreasing prices. There
+ * should always be a next schedule, but between rounds, liveSchedule is null.
+ *
+ * The lock period that the liquidators use might start before the previous
+ * round has finished, so we need to schedule the next round each time an
+ * auction starts. This means if the scheduling parameters change, it'll be a
+ * full cycle before we switch. Otherwise, the vaults wouldn't know when to
+ * start their lock period.
+ */
+
+const makeCancelToken = () => {
+  let tokenCount = 1;
+  return Far(`cancelToken${(tokenCount += 1)}`, {});
+};
+
+/**
+ * @typedef {object} AuctionDriver
+ * @property {() => void} reducePriceAndTrade
+ * @property {() => void} finalize
+ * @property {() => void} startRound
+ */
+
+/**
+ * @param {AuctionDriver} auctionDriver
+ * @param {import('@agoric/time/src/types').TimerService} timer
+ * @param {Awaited<import('./params.js').AuctionParamManaager>} params
+ * @param {import('@agoric/time/src/types').TimerBrand} timerBrand
+ */
+export const makeScheduler = async (
+  auctionDriver,
+  timer,
+  params,
+  timerBrand,
+) => {
+  // live version is non-null when an auction is active.
+  let liveSchedule;
+  // Next should always be defined after initialization unless it's paused
+  let nextSchedule;
+  const stepCancelToken = makeCancelToken();
+
+  // XXX why can't it be @type {AuctionState}?
```

Comment:
`ActionState` is an object, not a type. So it has to be,
```js
  /** @type {typeof AuctionState[keyof typeof AuctionState]} */
```

We are lacking a good enum facility. https://github.com/rbuckton/proposal-enum